### PR TITLE
Add an implementation of frames

### PR DESCRIPTION
**Pull request by ayberkt**

### CODE REVIEW

**Comment by ayberkt** [comment]
This PR is a draft (forgot to mark it as a draft initially and I am not sure how to declare it as such after it has been created). Also, I'm not sure why commits from #293 are listed here as I branched from 5e6b72a7685452c8df7ae386af124ba8ed3f019f. Did I mess something up?

In any case, I added some notation for working with families in `Foundations.Family` (f0f22ab8e75fb141c4d93651a43283d6cf82f8f5). I couldn't find something like this in the library but it might be that I missed it so I wanted to check (CC: @mortberg).

**Comment by ayberkt** [comment]
> (forgot to mark it as a draft initially and I am not sure how to declare it as such after it has been created)

Done now.

**Comment by ayberkt** [comment]
> Also, I'm not sure why commits from #293 are listed here as I branched from [5e6b72a](https://github.com/agda/cubical/commit/5e6b72a7685452c8df7ae386af124ba8ed3f019f).

I suppose that's because the commits were squashed and I branched from the `master` of my fork :facepalm:.

File: Cubical/Foundations/Family.agda [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+open import Cubical.Foundations.Logic
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+
+private
+  variable
+    ℓ₀ ℓ₁ ℓ₂ : Level
+
+Fam : (ℓ₀ : Level) → Type ℓ₁ → Type (ℓ-max (ℓ-suc ℓ₀) ℓ₁)
+Fam ℓ A = Σ (Set ℓ) (λ I → I → A)
```
Comment (mortberg) [inline Cubical/Foundations/Family.agda:14]: Please use the nice notation for Sigma types and rename all uses of Set to Type. Also, please don't use `I` for variables as we use it for the interval

Comment (mortberg) [reply Cubical/Foundations/Family.agda:14]: I guess you can use Fam in https://github.com/agda/cubical/blob/master/Cubical/Functions/Fibration.agda#L60

File: Cubical/Foundations/Family.agda [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+index (I , _) = I
+
+-- Application of a family over X to an index.
+_$_ : {A : Type ℓ₀} → (ℱ : Fam ℓ₂ A) → index ℱ → A
+_$_ (_ , f) = f
+
+infixr 7 _$_
+
+-- Membership for families.
+_ε_ : {A : Type ℓ₀} → A → Fam ℓ₁ A → Type _
```
Comment (mortberg) [inline Cubical/Foundations/Family.agda:26]: Why not \in instead

File: Cubical/Foundations/Family.agda [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+_⟨$⟩_ : {X : Type ℓ₀} {Y : Type ℓ₁} → (g : X → Y) → (ℱ : Fam ℓ₂ X) → Fam ℓ₂ Y
+g ⟨$⟩ ℱ = (index ℱ) , λ x → g (ℱ $ x)
+
+fmap : {X : Type ℓ₀} {Y : Type ℓ₁} → (g : X → Y) → (ℱ : Fam ℓ₂ X) → Fam ℓ₂ Y
+fmap = _⟨$⟩_
+
+syntax fmap (λ x → e) ℱ = ⁅ e ∣ x ε ℱ ⁆
+
+compri : {X : Type ℓ₀} → (I : Type ℓ₂) → (I → X) → Fam ℓ₂ X
+compri I f = (I , f)
```
Comment (mortberg) [inline Cubical/Foundations/Family.agda:39]: Does pairing really need another name?

File: Cubical/Foundations/Family.agda [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+-- Membership for families.
+_ε_ : {A : Type ℓ₀} → A → Fam ℓ₁ A → Type _
+x ε (_ , f) = fiber f x
+
+-- Composition of a family with a function.
+_⟨$⟩_ : {X : Type ℓ₀} {Y : Type ℓ₁} → (g : X → Y) → (ℱ : Fam ℓ₂ X) → Fam ℓ₂ Y
+g ⟨$⟩ ℱ = (index ℱ) , λ x → g (ℱ $ x)
+
+fmap : {X : Type ℓ₀} {Y : Type ℓ₁} → (g : X → Y) → (ℱ : Fam ℓ₂ X) → Fam ℓ₂ Y
```
Comment (mortberg) [inline Cubical/Foundations/Family.agda:33]: Is it really necessary to have two names for this?

File: Cubical/Foundations/Prelude.agda [hunk 5]
```diff
@@ -201,6 +201,12 @@ substRefl px = transportRefl px
 funExt : {f g : (x : A) → B x} → ((x : A) → f x ≡ g x) → f ≡ g
 funExt p i x = p x i
 
+funExt2 : {A₀ : Type ℓ₀} {A₁ : Type ℓ₁} {B : A₀ → A₁ → Type ℓ₂}
```
Comment (mortberg) [inline Cubical/Foundations/Prelude.agda:204]: This is already defined here: https://github.com/agda/cubical/blob/master/Cubical/Functions/FunExtEquiv.agda#L38

Maybe you can add a comment pointing to this file here?

File: Cubical/Structures/Frame.agda [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+open import Cubical.Foundations.Logic
+open import Cubical.Foundations.Family
+
+-- We will adopt the convention of referring using ℓ₀, ℓ₁, ℓ₂ for the
+-- carrier level, relation level, and the index type level respectively
+private
+  variable
+    ℓ₀ ℓ₁ ℓ₂ : Level
+
+-- TODO: is this defined somewhere?
```
Comment (mortberg) [inline Cubical/Structures/Frame.agda:21]: Yes: https://github.com/agda/cubical/blob/master/Cubical/Foundations/Function.agda#L41

File: Cubical/Structures/Frame.agda [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  isTop P ⊤ ⊓ isGLB P _∧_ ⊓ isLUB P ⋁_ ⊓ isDist P _∧_ ⋁_
+  where
+    P : Poset ℓ₀ ℓ₁
+    P = A , s
+
+FrameStr : (ℓ₁ ℓ₂ : Level) → Type ℓ₀ → Type _
+FrameStr ℓ₁ ℓ₂ = add-to-structure (RawFrameStr ℓ₁ ℓ₂) λ _ RF → [ FrameAx RF ]
+
+Frame : (ℓ₀ ℓ₁ ℓ₂ : Level) → Type _
+Frame ℓ₀ ℓ₁ ℓ₂ = Σ (Type ℓ₀) (FrameStr ℓ₁ ℓ₂)
```
Comment (mortberg) [inline Cubical/Structures/Frame.agda:91]: I guess you could use TypeWithStr here: https://github.com/agda/cubical/blob/master/Cubical/Foundations/Structure.agda#L14

File: Cubical/Structures/Frame.agda [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+RF-iso : {ℓ₁ ℓ₂ : Level} (M N : Σ (Type ℓ₀) (RawFrameStr ℓ₁ ℓ₂))
+       → fst M ≃ fst N → Type _
+RF-iso {ℓ₂ = ℓ₂} (A , (P , _) , ⊤₀ , _⊓₀_ , ⋁₀) (B , (Q , _), ⊤₁ , _⊓₁_ , ⋁₁) i =
+    (order-iso (A , P) (B , Q) i)
+  × (f ⊤₀ ≡ ⊤₁)
+  × ((x y : A) → f (x ⊓₀ y) ≡ (f x) ⊓₁ (f y))
+  × ((U : Fam ℓ₂ A) → f (⋁₀ U) ≡ ⋁₁ (f ⟨$⟩ U))
+  where
+    f = equivFun i
```
Comment (mortberg) [inline Cubical/Structures/Frame.agda:272]: Why not pattern-match on f?

File: Cubical/Structures/Frame.agda [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+            ϑ x y =
+              subst (λ { (_ , _ , _-_ , _) → 𝒻 (x - y) ≡ (𝒻 x) ⊓₁ (𝒻 y) }) (sym eq) refl
+
+            ξ : (U : Fam ℓ₂ A) → 𝒻 (⋁₀ U) ≡ ⋁₁ (index U , λ i → 𝒻 (U $ i))
+            ξ U = subst (λ { (_ , _ , _ , -) → 𝒻 (- U) ≡ ⋁₁ (𝒻 ⟨$⟩ U) }) (sym eq) refl
+
+        str-set : isSet (RawFrameStr ℓ₁ ℓ₂ A)
+        str-set = isSetΣ (PosetStr-set ℓ₁ A) λ _ →
+                  isSetΣ A-set₀ λ _ →
+                  isSetΣ (isSetΠ λ _ → isSetΠ λ _ → A-set₀) λ _ → isSetΠ λ _ → A-set₀
```
Comment (mortberg) [inline Cubical/Structures/Frame.agda:342]: Use isSetPi2. Is there isSetSigma3 or isSetx3?

File: Cubical/Structures/Frame.agda [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+
+        ret : (eq : F ≡ G) → f (g eq) ≡ eq
+        ret eq = str-set F G (f (g eq)) eq
+
+        RF-iso-prop : isProp (RF-iso (A , F) (A , G) (idEquiv A))
+        RF-iso-prop =
+          isPropΣ (RP-iso-prop (A , fst P) (A , fst Q) (idEquiv A)) (λ _ →
```
Comment (mortberg) [inline Cubical/Structures/Frame.agda:349]: Use the lemmas with higher numbers

**Comment by ayberkt** [comment]
@mortberg thanks for the feedback. I have made some substantial changes to posets in my own development so I will first send that as a PR before completing this.

**Comment by ayberkt** [comment]
@mortberg also, a technical question: the notion of a σ-frame is the special case of the notion of a frame (as defined [here](https://ayberkt.gitlab.io/msc-thesis/Frame.html#3229)) in which the index set is `ℕ`. I am not sure what would be a good way of expressing this state of affairs in Agda would be.

Any suggestions?

**Comment by mortberg** [comment]
> @mortberg also, a technical question: the notion of a σ-frame is the special case of the notion of a frame (as defined [here](https://ayberkt.gitlab.io/msc-thesis/Frame.html#3229)) in which the index set is `ℕ`. I am not sure what would be a good way of expressing this state of affairs in Agda would be.
> 
> Any suggestions?

Hmm, I'm not sure what the problem is. Can't you just use `FrameStr ℕ`?

**Comment by ayberkt** [comment]
`FrameStr` is parametrised by the carrier set so `FrameStr ℕ` is merely an arbitrary frame on the carrier set `ℕ`, i.e., it has arbitrary joins and not just countable ones.

Parametrising my notion of frame by an index set will not work as I must have that any family over the carrier set has a join, regardless of what the index set of the family is. So I need to somehow restrict `FrameStr` by a condition to be imposed on the index sets of the families, and this doesn't sound like a good solution to me.

**Comment by ayberkt** [comment]
Actually, I am also confused by the wording of the HoTT book: "a lattice with countable joins in
which binary meets distribute over countable joins".

It seems to me that they are abusing language by meaning _countably infinite_ with "countable". At least, this is what is implemented by Niccolò's slides (as implemented in #316).

**Comment by mortberg** [comment]
> Actually, I am also confused by the wording of the HoTT book: "a lattice with countable joins in
> which binary meets distribute over countable joins" (pg. 375).
> 
> It seems to me that they are abusing language by meaning _countably infinite_ with "countable". At least, this is what is defined by Niccolò's slides (as implemented in #316).

I think it's pretty common to use the term "countable" for "countably infinite" alone (see https://en.wikipedia.org/wiki/Countable_set). I'm not sure what they mean by the word in the HoTT book. Maybe open an issue about it for the book?

**Comment by ayberkt** [comment]
I never got around to finishing this and now too many things have changed that I think it will be easier to do this from scratch. I will make a new PR at some point, taking into account the feedback given here.

**Pull request closed without merging**